File: packages/core/src/programmers/helpers/AtomicPredicator.ts
import { Atomic } from "@typia/interface";
import { ArrayUtil } from "@typia/utils";

import { MetadataSchema } from "../../schemas/metadata/MetadataSchema";

export namespace AtomicPredicator {
  export const constant = (props: {
    metadata: MetadataSchema;
    name: Atomic.Literal;
  }): boolean =>
    !ArrayUtil.has(
      props.metadata.natives,
      (native) => native.name.toLowerCase() === props.name,
    );

  export const atomic = (props: {
    metadata: MetadataSchema;
    name: Atomic.Literal;
  }): boolean =>
    !ArrayUtil.has(
      props.metadata.natives,
      (native) => native.name.toLowerCase() === props.name,
    );

  export const native = (name: string) => LIKE.has(name.toLowerCase());

  export const template = (metadata: MetadataSchema): boolean =>
    !ArrayUtil.has(
      metadata.natives,
      (native) => native.name.toLowerCase() === "string",
    );
}

const LIKE = new Set(["boolean", "bigint", "number", "string"]);

File: packages/core/src/programmers/helpers/AtomicPredicator.go
package helpers

type atomicPredicatorNamespace struct{}

var AtomicPredicator = atomicPredicatorNamespace{}

func (atomicPredicatorNamespace) Constant(name string, metadata Metadata) bool {
	for _, constant := range GetSlice(metadata, "constants") {
		if GetName(constant) == name {
			return true
		}
	}
	return false
}

func (atomicPredicatorNamespace) Atomic(name string, metadata Metadata) bool {
	for _, atomic := range GetSlice(metadata, "atomics") {
		if GetName(atomic) == name {
			return true
		}
	}
	return false
}

func (atomicPredicatorNamespace) Native(name string, metadata Metadata) bool {
	for _, native := range GetSlice(metadata, "natives") {
		if GetName(native) == name {
			return true
		}
	}
	return false
}

func (atomicPredicatorNamespace) Template(metadata Metadata) bool {
	return len(GetSlice(metadata, "templates")) != 0
}
